### PR TITLE
Add a safety guard to Wilcoxon pruner, and modify the docstring

### PR DESCRIPTION
**Pull request by eukaryo**
## Motivation

I want to make Wilcoxon pruner more safe and effective.

## Description of the changes

- If the value of the best_trial was not better than the average of the current trial's intermediate_values, the pruner returns False.
- Docstring is modified. In each trial, the modified example code shuffles the order in which data is processed. It never shuffles the dataset itself.

### CODE REVIEW

**Comment by nabenabe0928** [comment]
@y0z @contramundum53 Could you review this PR?

File: optuna/pruners/_wilcoxon.py [hunk 1]
```diff
@@ -58,22 +58,22 @@ def eval_func(param, input_):
...
+                ordering = rng.permutation(range(len(input_data)))
+                s = []
+                for i in ordering:
                     param = trial.suggest_float("param", -1, 1)
                     loss = eval_func(param, input_data[i])
                     trial.report(loss, i)
-                    s += loss
+                    s.append(loss)
                     if trial.should_prune():
                         raise optuna.TrialPruned()
```
Comment (contramundum53) [inline optuna/pruners/_wilcoxon.py:74]:
```suggestion
                        return sum(s) / len(s)
```
Could you also change the docstring to explain that this is necessary, because this is different from other pruners?

File: optuna/pruners/_wilcoxon.py [hunk 2]
```diff
@@ -167,6 +169,21 @@ def prune(self, study: "optuna.study.Study", trial: FrozenTrial) -> bool:
 
         alt = "less" if study.direction == StudyDirection.MAXIMIZE else "greater"
 
+        if alt == "less":
+            average_is_best = best_trial.value <= sum(step_values) / len(step_values)
+        else:
+            average_is_best = best_trial.value >= sum(step_values) / len(step_values)
```
Comment (contramundum53) [inline optuna/pruners/_wilcoxon.py:175]:
```suggestion
        if study.direction == StudyDirection.MAXIMIZE:
            alt = "less"
            average_is_best = best_trial.value <= sum(step_values) / len(step_values)
        else:
            alt = "greater"
            average_is_best = best_trial.value >= sum(step_values) / len(step_values)
```
A nit.

**Review by contramundum53 — APPROVED** [review]
Otherwise, LGTM.

File: optuna/pruners/_wilcoxon.py [hunk 3]
```diff
@@ -167,6 +169,21 @@ def prune(self, study: "optuna.study.Study", trial: FrozenTrial) -> bool:
...
+
         # We use zsplit to avoid the problem when all values are zero.
         p = ss.wilcoxon(diff_values, alternative=alt, zero_method="zsplit").pvalue
+
+        if p < self._p_threshold and average_is_best:
+            # WilcoxonPruner found the current trial should be pruned,
+            # but the value of the best trial was not better than
+            # the average of the current trial's intermediate values.
+            # For safety, WilcoxonPruner concludes it should not be pruned.
+            # If this if-statement is frequently satisfied,
```
Comment (y0z) [inline optuna/pruners/_wilcoxon.py:185]: Although this is a simple question, is there a reasonable way that users can detect `this if-statement is frequently satisfied`?

File: optuna/pruners/_wilcoxon.py (hunk 1 above)
Comment (eukaryo) [reply optuna/pruners/_wilcoxon.py:75]: I agree that the suggested technique is a beneficial workaround. However, I believe it might be too advanced for use in a docstring example. In addition, it is not necessary; for example, it has no effect with GridSampler.
I added an comment in the code and wrote a separate note.

Comment (contramundum53) [reply optuna/pruners/_wilcoxon.py:75]: If you add a note in the docstring, why not demonstrate that in the example?
This workaround is actually necessary for all samplers except for `GridSampler`, `BruteForceSampler`, `QMCSampler` and `RandomSampler` I think.

Comment (eukaryo) [reply optuna/pruners/_wilcoxon.py]: I see; I modified the example so that `return sum(s) / len(s)`.

**Review by y0z — APPROVED** [review]
LGTM

**Review by contramundum53 — APPROVED** [review]
LGTM.
